design/resource-validation.md

```diff
@@ -0,0 +1,217 @@
+  corresponding [location](https://nginx.org/en/docs/http/ngx_http_core_module.html#location) block.
+* The rules don't cover unsupported field cases. For example, the webhook does not know which filters are implemented by
+  NKG, thus it cannot generate an appropriate error for NKG.
+
+Additionally, as mentioned in [GEP-922](https://gateway-api.sigs.k8s.io/geps/gep-922/#implementers),
+"implementers must not rely on webhook or CRD validation as a security mechanism. If field values need to be escaped to
+secure an implementation, both webhook and CRD validation can be bypassed and cannot be relied on."
+
+## Requirements
+
+Design a validation mechanism for Gateway API resources.
+
+### Personas
+
+* *Cluster admin* who installs Gateway API (the CRDs and Webhook), installs NKG, creates Gateway and GatewayClass
+  resources.
+* *Application developer* who creates HTTPRoutes and other routes.
+
+### User Stories
+
+1. As a cluster admin, I'd like to share NKG among multiple application developers, specifically in a way that invalid
+   resources of one developer do not affect on the resources of the other developers.
+2. As a cluster admin/application developer, I expect that NKG rejects any invalid resources I create and I am able to
+   see the reasons (errors) for that.
+
+### Goals
+
+* Ensure that NKG continues to work and/or fails predictably in the face of invalid input.
+* Ensure that both cluster admin and application developers can see the validation errors reported about the resource
+  they create (own).
+* For the best UX, minimize the feedback loop: users should be able to see most of the validation errors reported by a
+  Kubernetes API server during a CRUD operation on a resource.
+* Ensure that the validation mechanism conforms to the Gateway API spec.
+
+### Non-Goals
+
+* Validation of non-Gateway API resources: Secrets, EndpointSlices. (For example, a TLS Secret resource might include a
+  non-valid TLS cert that will make NGINX fail to reload).
+
+## Design
+
+We will introduce two validation methods to be run by NKG control plane:
+
+1. Re-run of the Gateway API webhook validation
+2. NKG-specific field validation
+
+### Re-run of Webhook Validation
+
+Before processing a resource, NKG will validate it using the functions from
+the [validation package](https://github.com/kubernetes-sigs/gateway-api/tree/b241afc88e68c952cc0a59a5c72a51358dc2bada/apis/v1beta1/validation)
+from the Gateway API. This will ensure that the webhook validation cannot be bypassed (it can be bypassed if the webhook
+is not installed, misconfigured, or running a different version), and it will allow us to avoid repeating the same
+validation in our code.
+
+If a resource is invalid:
+
+* NKG will not process it -- it will treat it as if the resource didn't exist. This also means that if the resource was
+  updated from a valid to an invalid state, NKG will also ignore any previous valid state. For example, it will remove
```

Comment:
This is an interesting caveat. Is this standard behavior in NIC and other ingress/gateway implementations?

It's different than the behavior of nginx and Kubernetes update failures, where the previous valid config is maintained.

Reply:
The idea here is to reconcile to the latest state, which is used extensively in Kubernetes. Invalid state, if it is stored in the etcd (it passed CRD and Webhook validation rules) is still the latest state. This way, in case of control plane restarts / upgrades, the regenerated configuration will still be the same. Otherwise, the restarted/upgraded control plane will not be able to see the latest "valid" state.

NIC does it similarly -- it doesn't have a webhook.

Other Gateway implementations I've looked do not re-run webhook validation -- they do their own validation while parsing the resources. At the same time, they will not include the resource configuration into the data plane configuration for certain cases, which is similar to rejecting the resource.

Reply:
Makes sense. Thanks for the explanation.